### DISCUSSION TITLE
Adding table entry

### DISCUSSION THREAD

**Discussion by g83wang** [NIKSS-vSwitch/nikss#102] [Q&A] [answered]
Hi Team, 

I am trying to manage a table of the p4c-ebpf-psa-upf program below by using the nikss commands.
https://github.com/p4lang/p4c/blob/main/backends/ebpf/psa/examples/upf.p4

The table is defined as below in the upf example:

table ingress_l4_src_port {
        actions = {
            nop;
            set_ingress_src_port_range_id;
        }
        key = {
            meta.upf.l4_sport: exact; // range macthing unsupported
        }
        size = 512;
    }

I attached the program to pipeline 0:
~/p4c$ sudo nikss-ctl pipeline show id 0
{
    "pipeline": {
        "id": 0,
        "load_time": "2023-06-05T10:36:36-0400",
        "bpf_hook": "TC",
        "has_egress_program": false,
        "ports": [],
        "objects": [
            "ingress_upf_process_ingress_l4port_ingress_l4_src_port",
            "ingress_upf_process_ingress_l4port_ingress_l4_dst_port",
            "ingress_upf_ingress_source_interface_lookup_by_port",
            "ingress_upf_ingress_session_lookup_by_ue_ip",
            "ingress_upf_ingress_session_lookup_by_teid",
            "ingress_upf_ingress_pdr_lookup",
            "ingress_upf_ingress_far_lookup",
            "ingress_ip_forward_ipv4_lpm",
            "ingress_upf_process_ingress_l4port_ingress_l4port_fields"
        ]
    }

If I try to add a table entry by using the command here:
sudo nikss-ctl table add pipe 0 ingress_upf_process_ingress_l4port_ingress_l4_src_port action _nop key 5353

I will get the error "_nop: unknown action specification"

I think I am following the naming convention below correctly?  
https://github.com/NIKSS-vSwitch/nikss/tree/master#naming-convention

Could I get some ideas on this issue?
Also, is there any documentation on the commands I can read other than the command reference?

Thank you.

**Comment by g83wang** [comment]
Resolved. Thanks

**Comment by osinstom** [comment]
@g83wang Can you post the solution here for future reference?

**Reply by g83wang** [reply, marked as answer]
Yes. 

When specifying the action by name, I forgot to add the string "name" in the command.

If we look into the source code, after we match the "action", nikss will look for either "id" or "name":

            else if (is_keyword(**argv, "action")) {
                    NEXT_ARGP_RET();
            
                    if (is_keyword(**argv, "id")) {
                        NEXT_ARGP_RET();
                        char *ptr = NULL;
                        nikss_action_set_id(action, strtoul(**argv, &ptr, 0));
                        if (*ptr) {
                            fprintf(stderr, "%s: unable to parse as an action id\n", **argv);
                            return EINVAL;
                        }
                    } else if (is_keyword(**argv, "name")) {
                        NEXT_ARGP_RET();
                        uint32_t action_id = nikss_table_get_action_id_by_name(ctx, **argv);
                        if (action_id == NIKSS_INVALID_ACTION_ID) {
                            fprintf(stderr, "%s: action not found\n", **argv);
                            return EINVAL;
                        }

And I also found this in the doc  (https://github.com/NIKSS-vSwitch/nikss/blob/master/docs/command%20reference.md):

("ACTION := { id ACTION_ID | name ACTION_NAME }")

I directly jumped in nikss from p4c. Should have read the nikss doc carefully. My bad. Sorry.
